core/crypto/nike/hybrid/hybrid.go

```diff
@@ -0,0 +1,210 @@
+}
+
+func (s *scheme) DerivePublicKey(privKey nike.PrivateKey) nike.PublicKey {
+	return &publicKey{
+		scheme: s,
+		first:  privKey.(*privateKey).scheme.first.DerivePublicKey(privKey.(*privateKey).first),
+		second: privKey.(*privateKey).scheme.second.DerivePublicKey(privKey.(*privateKey).second),
+	}
+}
+
+func (s *scheme) Blind(groupMember []byte, blindingFactor []byte) (blindedGroupMember []byte) {
+	if len(groupMember) != s.PublicKeySize() {
+		panic("invalid group member size")
+	}
+	if len(blindingFactor) != s.PrivateKeySize() {
+		panic("invalid blinding factor size")
+	}
+	return append(s.first.Blind(groupMember[:s.first.PublicKeySize()], blindingFactor[:s.first.PrivateKeySize()]),
+		s.second.Blind(groupMember[s.first.PublicKeySize():], blindingFactor[s.first.PrivateKeySize():])...)
+}
+
+func (s *scheme) NewEmptyPublicKey() nike.PublicKey {
+	return &publicKey{
+		scheme: s,
+		first:  s.first.NewEmptyPublicKey(),
+		second: s.second.NewEmptyPublicKey(),
+	}
+}
+
+func (s *scheme) NewEmptyPrivateKey() nike.PrivateKey {
+	return &privateKey{
+		scheme: s,
+		first:  s.first.NewEmptyPrivateKey(),
+		second: s.second.NewEmptyPrivateKey(),
+	}
+}
+
+func (s *scheme) UnmarshalBinaryPublicKey(b []byte) (nike.PublicKey, error) {
+	pubkey := s.NewEmptyPublicKey()
+	err := pubkey.FromBytes(b)
+	if err != nil {
+		return nil, err
+	}
+	return pubkey, nil
+}
+
+func (p *privateKey) Reset() {
+	p.first.Reset()
+	p.second.Reset()
+}
+
+func (p *privateKey) Bytes() []byte {
+	return append(p.first.Bytes(), p.second.Bytes()...)
+}
+
+func (p *privateKey) FromBytes(b []byte) error {
+	err := p.first.FromBytes(b[:p.scheme.first.PrivateKeySize()])
+	if err != nil {
+		return err
+	}
+	return p.second.FromBytes(b[p.scheme.first.PrivateKeySize():])
+}
+
+// MarshalBinary is an implementation of a method on the
+// BinaryMarshaler interface defined in https://golang.org/pkg/encoding/
+func (p *privateKey) MarshalBinary() ([]byte, error) {
+	return p.Bytes(), nil
+}
+
+// UnmarshalBinary is an implementation of a method on the
+// BinaryUnmarshaler interface defined in https://golang.org/pkg/encoding/
+func (p *privateKey) UnmarshalBinary(data []byte) error {
+	return p.FromBytes(data)
+}
+
+// MarshalText is an implementation of a method on the
+// TextMarshaler interface defined in https://golang.org/pkg/encoding/
+func (p *privateKey) MarshalText() ([]byte, error) {
+	return []byte(base64.StdEncoding.EncodeToString(p.Bytes())), nil
+}
+
+// UnmarshalText is an implementation of a method on the
+// TextUnmarshaler interface defined in https://golang.org/pkg/encoding/
+func (p *privateKey) UnmarshalText(data []byte) error {
+	raw, err := base64.StdEncoding.DecodeString(string(data))
+	if err != nil {
+		return err
+	}
+	return p.FromBytes(raw)
+}
+
+func (p *publicKey) Blind(blindingFactor []byte) error {
+	err := p.first.Blind(blindingFactor[:p.scheme.first.PublicKeySize()])
+	if err != nil {
```

Comment:
Since we may have written stuff to `p.first` in these methods, it would perhaps be safer if we called `p.Reset()` before returning `err`, to prevent people from using partially overwritten things?

Reply:
yeah sounds good

Reply:
done